Enable ability to write JS app _mostly_ in Wasm'd Dark

StachuDotNet opened the pull request:
Changelog:

```
Area of change
- details
```

The goal of this PR has been to provide `dark-editor`/`experiment1` a way to move forward without maintaining all its state in JS, and generally achieve the "write as much of the client as possible in Dark" virtue, as we don't have a good way of doing that currently.

---

Here's a small demo to that end:

https://user-images.githubusercontent.com/906686/236076699-89140418-84e3-4688-bb1f-84068352ed4c.mp4

These are all handled in Dark:
- defining the type of the global 'state' of the app
- modeling the Events that can occur in the app
- maintaining an app-specific `state`
- listening to events reported by the UI (JS), and updating state based on them
- informing the UI of state-updates

These are handled in JS/Vue
- rendering
- listening for UI events
- informing the WASM runtime/program of UI events
- listening to WASM for state updates

This implementation has leaned in the "Elm-ish" direction, but that's not necessary and could be reverted.

---

To support this, I've:
- exposed a small # of stdlib functions (i.e. `WASM.setState`, `WASM.getState`, `WASM.callJSFunction`) to manage a program, and its state, in a WASM-run Dark program
- added a `wasm.dark` file to represent the "dark program which runs _in the wasm runtime_" (collection of types and fns, some of which may call out to JS)
- added additional methods on the `darklang` object exposed to JS, to _load_ that program, and call functions within it (currently there's `darklang.evalExpr` which we've been using - this adds `darklang.loadProgram`, `darklang.callUserFunction`)
- added a demo canvas `dark-js-wasm-io` which demos a small Vue app which uses this functionality (all state is managed "in Dark" and rendering + event-listening is handled "in JS")

---

the Dark side of things is handled in a `wasm.dark` file:
```fsharp
type RawEvent =
  { eventName: String
    data: String }

type Event =
  | Increment
  | Decrement
  | Set of Int

let parseEvent (rawEvent: RawEvent): Result<Event, String> =
  match rawEvent.eventName with
  | "increment" -> Ok Event.Increment
  | "decrement" -> Ok Event.Decrement
  | "reset" -> Ok Event.Reset
  | "setCounter" ->
    match Json.parse<Int> rawEvent.data with
    | Ok i -> Ok (Event.Set i)
    | Error err -> Error "Couldn't parse new counter value"
  | _ -> Error "Couldn't parse event name"


let init (): Result<Int, String> =
  WASM.setState<Int> 0

let handleEvent (evt: RawEvent): Result<Unit, String> =
  match parseEvent evt with
  | Ok evt ->
    match WASM.getState<Int> with
    | Ok currentState ->
      let newState =
        match evt with
        | Increment -> currentState + 1
        | Decrement -> currentState - 1
        | Set i -> i

      // returns result, but let's assume it worked...
      let _ = WASM.setState<Int> newState

      match Json.serialize<Int> newState with
      | Ok serialized -> WASM.callJSFunction "window.stateUpdated" [serialized]
      | Error err -> Error "Couldn't serialize state"
    | Error err -> Error "Couldn't get state"
  | Error err -> Error "Couldn't parse event from raw event"
```


the JS/Vue side of things handles events/state like this:
```js
const app = new Vue({
  el: '#app',
  data: { counter: 0 },
  methods: {
    // handle state updated by WASM
    updateState(newState) { this.counter = newState; },

    // something happened in the UI - tell the WASM'd Dark program!
    async eventHappened(event) {
      // this is hacky!
      const typeOfArg =
        JSON.stringify(
          JSON.stringify(
            [ "TCustomType", [ "User", { "modules": [], "typ": "RawEvent", "version": 0 } ], []]
          )
        );
      const arg = JSON.stringify(JSON.stringify(event));
      const _handleEventResult = await parseAndEval(darklang, `WASM.callUserFunction "handleEvent" [${typeOfArg}] [${arg}]`);
    },
  },
  mounted() {
    window.stateUpdated = (newState) => { this.updateState(JSON.parse(newState)); };
  },
});
```

---

This solution has the shape of something useful, but feels a bit wrong still. Would appreciate input.

Some misgivings:
- having a `WASM.callJSFunction` function available feels potentially dangerous
- _how_ we load that program longer-term may likely change - maybe the wasm should _fetch_ the program (types+fns) instead? Once we have a stable package manager, _package_ fns and types are likely to be loaded into the WASM runtime _not_ from the JS host (i.e. by streaming/fetching)
- there's some ugliness in calling Dark user functions from JS

pbiggar commented on backend/static/tree-sitter/tree-sitter-darklang.wasm:
Dunno if you intended to check .wasm files in

StachuDotNet commented:
I feel like I haven't been clear what I'm working towards, so:
The goal of this PR has been to provide `dark-editor`/`experiment1` a way to move forward without maintaining all its state in JS, and generally achieve the "write as much of the client as possible in Dark" virtue, as we don't have a good way of doing that currently.

To that end, I've
- been exposing a small # of stdlib functions (i.e. `WASM.setState`, `WASM.getState`, `WASM.callJSFunction`)
- added a `wasm.dark` file to represent the "dark program which runs _in the wasm runtime_" (collection of types and fns, some of which may call out to JS)
- added additional methods on the `darklang` object exposed to JS, to _load_ that program, and call functions within it (currently there's `darklang.evalExpr` which we've been using - this adds `darklang.loadProgram`, `darklang.callUserFunction`)
- added a demo canvas `dark-js-wasm-io` which demos a small Vue app which uses this functionality (all state is managed "in Dark" and rendering + event-listening is handled "in JS")

How we load that program longer-term may likely change - maybe the wasm should _fetch_ the program (types+fns) instead?

Anyway - I should have been clearer earlier about this, but think it's a step in the right direction. in any case, along with the HttpClient being usable within WASM, and packages being available, it feels like these efforts will combine well and allow us to focus more on the actual application.

Close to wrapping this up, but wanted to check in. Just a few minor technical tasks left before tidying this up and requesting review.

StachuDotNet commented:
https://user-images.githubusercontent.com/906686/236076699-89140418-84e3-4688-bb1f-84068352ed4c.mp4

These are all handled in Dark:
- defining the type of the global 'state' of the app
- the Events that occur
- updating state based on events
- informing the UI of state-updates

These are handled in JS/Vue
- listening for events
- rendering
- informing the WASM runtime/program of events

the Dark side of things is handled in a `wasm.dark` file:
```fsharp
type RawEvent =
  { eventName: String
    data: String }

type Event =
  | Increment
  | Decrement
  | Set of Int

let parseEvent (rawEvent: RawEvent): Result<Event, String> =
  match rawEvent.eventName with
  | "increment" -> Ok Event.Increment
  | "decrement" -> Ok Event.Decrement
  | "reset" -> Ok Event.Reset
  | "setCounter" ->
    match Json.parse<Int> rawEvent.data with
    | Ok i -> Ok (Event.Set i)
    | Error err -> Error "Couldn't parse new counter value"
  | _ -> Error "Couldn't parse event name"


let init (): Result<Int, String> =
  WASM.setState<Int> 0

let handleEvent (evt: RawEvent): Result<Unit, String> =
  match parseEvent evt with
  | Ok evt ->
    match WASM.getState<Int> with
    | Ok currentState ->
      let newState =
        match evt with
        | Increment -> currentState + 1
        | Decrement -> currentState - 1
        | Set i -> i

      // returns result, but let's assume it worked...
      let _ = WASM.setState<Int> newState

      match Json.serialize<Int> newState with
      | Ok serialized -> WASM.callJSFunction "window.stateUpdated" [serialized]
      | Error err -> Error "Couldn't serialize state"
    | Error err -> Error "Couldn't get state"
  | Error err -> Error "Couldn't parse event from raw event"
```


the JS/Vue side of things handles events like this:
```js
const app = new Vue({
  el: '#app',
  data: { counter: 0 },
  methods: {
    // handle state updated by WASM
    updateState(newState) { this.counter = newState; },

    // something happened in the UI - tell the WASM'd Dark program!
    async eventHappened(event) {
      // this is hacky!
      const typeOfArg =
        JSON.stringify(
          JSON.stringify(
            [ "TCustomType", [ "User", { "modules": [], "typ": "RawEvent", "version": 0 } ], []]
          )
        );
      const arg = JSON.stringify(JSON.stringify(event));
      const _handleEventResult = await parseAndEval(darklang, `WASM.callUserFunction "handleEvent" [${typeOfArg}] [${arg}]`);
    },
  },
  mounted() {
    window.stateUpdated = (newState) => { this.updateState(JSON.parse(newState)); };
  },
});
```
